server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -709,6 +739,21 @@ public DiscoveryNodes nodes() {
             return nodes;
         }
 
+        public Builder putTransportVersion(String node, TransportVersion version) {
+            transportVersions.put(node, Objects.requireNonNull(version, node));
+            return this;
+        }
+
+        public Builder transportVersions(Map<String, TransportVersion> versions) {
+            versions.forEach((key, value) -> Objects.requireNonNull(value, key));
+            this.transportVersions.putAll(versions);
+            return this;
+        }
+
+        public Map<String, TransportVersion> transportVersions() {
+            return new HashMap<>(this.transportVersions);
```

Comment:
Can this just be Collections.unmodifiableMap so that a copy is not made?

Reply:
The resulting map is modified in NodeJoinExecutor, but that copy should be done there.